lib/propolis/src/firmware/smbios/mod.rs

```diff
@@ -0,0 +1,198 @@
+    pub fn add(
+        &mut self,
+        handle: Handle,
+        table: &dyn Table,
+    ) -> Result<(), TableError> {
+        let table_bytes = table.render(handle);
+        let table_size = table_bytes.len();
+        assert!(table_size != 0, "table should not be zero-length");
+
+        if let Some(conflict) = self.tables.insert(handle, table_bytes) {
+            // replace the item which we conflicted with in the first place
+            let _ = self.tables.insert(handle, conflict);
+            Err(TableError::HandleConflict(handle))
+        } else {
+            self.total_size += table_size;
+            self.max_single_size = usize::max(self.max_single_size, table_size);
+            Ok(())
+        }
+    }
+
+    /// Build Entry Point structure.  Emits the raw-byte values of both the
+    /// Entry Point and the associated Structure Table data.
+    pub fn commit(self) -> TableBytes {
+        let mut data = bits::EntryPoint::new();
+        // hardcode to version 2.7 for now
+        data.major_version = 2;
+        data.minor_version = 7;
+        data.bcd_revision = 0x27;
+        data.table_length = self.total_size as u16;
+        data.num_structs = self.tables.len() as u16;
+        data.max_struct_sz = self.max_single_size as u16;
+        data.update_cksums();
+
+        let mut table_data = Vec::with_capacity(self.total_size);
+        for (handle, table) in self.tables.iter() {
+            // copy all non-EoT tables
+            if *handle != self.eot_handle {
+                table_data.extend_from_slice(table);
+            }
+        }
+        // end-of-table goes at the end
+        table_data.extend_from_slice(
+            self.tables.get(&self.eot_handle).expect("EoT entry is present"),
+        );
+
+        TableBytes {
+            entry_point: data.to_raw_bytes().to_vec(),
+            structure_table: table_data,
+        }
+    }
+}
+
+pub struct TableBytes {
+    pub entry_point: Vec<u8>,
+    pub structure_table: Vec<u8>,
+}
+
+/// Possible errors when adding [Table] entries to [Tables]
+#[derive(thiserror::Error, Debug)]
+pub enum TableError {
+    #[error("Conflicting handle {0}")]
+    HandleConflict(Handle),
+}
+
+/// Structure Handle
+///
+/// A 16-bit number used to uniquely identify a single structure in a collection
+/// of SMBIOS tables.
+///
+/// Defaults to 0xffff, which is considered "Unknown" in most SMBIOS tables.
+#[derive(Copy, Clone, Eq, PartialEq, Hash, Debug, Ord, PartialOrd)]
+#[repr(transparent)]
+pub struct Handle(pub u16);
+impl Handle {
+    pub const UNKNOWN: Self = Self(0xffff);
+}
+impl fmt::Display for Handle {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Display::fmt(&self.0, f)
+    }
+}
+/// Default to the Unknown handle
+impl Default for Handle {
+    fn default() -> Self {
+        Self::UNKNOWN
+    }
+}
+impl From<u16> for Handle {
+    fn from(value: u16) -> Self {
+        Self(value)
+    }
+}
```

Comment:
I wonder if we ought to have a constructor that asserts that a handle is less than 0xFF00, since I believe the range above that is reserved (per Table 3 in DSP0134 v3.7.0  § 6.1.2).

Reply:
But `0xffff` is "valid" as the unknown handle.  For now I'd like to leave this be.

Reply:
Yeah, I was thinking that if you wanted the "unknown" handle, you would refer to it through `Handle::default()` or `Handle::UNKNOWN` rather than `Handle::from(0xffff)`. but, it's not important...